src/NUnitFramework/framework/Internal/Execution/WorkItemQueue.cs

```diff
@@ -359,17 +365,19 @@ internal void Restore()
             // else can take a look at it later on.
             //Guard.OperationValid(State != WorkItemQueueState.Running, $"Attempted to restore state of {Name} while queue was running.");
 
-            var savedQueues = _savedState.Pop().InnerQueues;
+            var state = _savedState.Pop();
 
             // If there are any queued items, copy to the next lower level
             for (int i = 0; i < PRIORITY_LEVELS; i++)
             {
                 WorkItem work;
                 while (_innerQueues[i].TryDequeue(out work))
-                    savedQueues[i].Enqueue(work);
+                    state.InnerQueues[i].Enqueue(work);
             }
 
-            _innerQueues = savedQueues;
+            _innerQueues = state.InnerQueues;
+            _addId += state.AddId;
+            _removeId += state.RemoveId;
```

Comment:
These two lines bother me. As you say, @oznetmaster is the expert in this area, but everywhere else in the code we use `Interlocked.CompareExchange` to change these values, plus the `_addId` and `_removeId` start at `int.MinValue` and increment, so won't this subtract from the ids and then eventually start adding to them? That doesn't seem like it is restoring the values. How am I reading this wrong?

Reply:
If it's possible for anyone else to be changing the queues while we are saving or restoring the state, then we have lots bigger problems than the counts! We rely on the fact that the methods are only called when nothing else can change the queue state.

This reliance on being called correctly has bothered me as well. Can we place a Guard statement somewhere such that it will throw if the precondition is not satisfied?

As to the specific issue, I wanted the newly created levels to start at zero so I could add them into the previous level when restoring the state. An alternative approach would be to leave the values as they are - truly restore them - but that will only work if the queue we are restoring had completely balanced adds and removes. That is not always the case, which is why I copy any "left over" queue items. If you think it's better, I could simply count the number of items I copy and add those values to the `_addId`.

Bear in mind that the only purpose of these integers is to provide two monotonic increasing values as a unique signature for the state of the queue as well as an indication of whether the queue contains anything.

Reply:
Sounds good, thanks for the explanation. As I said before, I find this non-locking code to be overly complex and I wonder if it is truly necessary. One day, I would like to do a spike, switch it to locking code, see how much cleaner it is and also run some timings on large test suites.

The `_addId` and `_removeId` usages might be clearer if they were `uint` so they started at zero, but  `Interlocked.CompareExchange` doesn't have an overload for that 😦

Reply:
@rprouse Yes, non-locking code is more complex then locking code.  Look at one of the concurrent collection classes if you want to blow your mind.  However, the speed difference can really be significant. 

@CharliePoole I do not think that the ids ever need to be saved and restored.  They are really only used to provide the non-locking control over the queue, to make sure that some other thread has not gotten there before you.  As long as they are unique, and continue to increase, they should just work, no matter what level of queue stack they are used on.

Reply:
@CharliePoole I guess I am confused. "but that will only work if the queue we are restoring had completely balanced adds and removes.".  I am not sure why that would be the case?

However, if new items are "copied" to the restored queue as part of restoring it, then the _addId should already reflect those items having been previously added, and since they have not been removed, the _removeId should reflect that as well.

Reply:
This would also imply that the initialization of the two ids should be left at int.MinValue.

Reply:
@oznetmaster The reason for my confusing comment is that there is code in `WorkItemQueue` that uses the difference between adds and removes to determine if the queue has data. I don't know if it's your code or if somebody else added it, but that was the source of the error that this fixes. So I tried to make sure that adds and removes increase in a monotonic fashion __and__ that the difference between them always represents the number of items in the queue. Proof of the pudding is that it fixed the problem it was meant to fix.

Your comment starting "However..." is correct. That was my intent in adding in the adds and removes.

I use an initialization of zero on any isolated queues that are created so that the value can be meaningfully added to the original queue when we go back to it. That is, the original queue ends up with the values it would have had if we had continued to use it rather than creating an isolated queue. As you say, it's more complicated than a simple lock. 😄

Reply:
If you do not save and restore the two ids, and do the following, everything should work correctly.

You should replace 
```
cachedRemoveId == cachedAddId
```
with
```
IsEmpty
```

and then the relative values of the two ids will no longer matter, since they will not be used to determine whether the queue is empty.

I think doing it this way should be more reliable.  I have not actually tried it, however :(

Reply:
Yes, I didn't change that because I wasn't sure at the time how the ids worked and why you were making that comparison. Instead, since my introduction of save / restore had broken the add and remove ids, I decided to fix them so they would end up with the value they would have had without the save and restore.

If you see a problem with what is done, i.e. if it will cause problems, then it can either wait till I'm able to go back to it or somebody else can take over and redo it. Originally, the problem this solves appeared to be quite critical, but since nobody has been complaining, it may not have been. I wish I could have had your eyes on this when I was doing it! Right now it would take me a bit of effort to bring my mind back to this PR. Do you want to take it on?

Reply:
Since the environment I currently support NUnit in is with CF on a single core processor, I really have no way to adequately test any changes I might make.

I think someone else will need to take this on.

Reply:
Use of IsEmpty in the middle of the dequeue loop causes an exception. See my further comment below.